aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1601,6 +1601,154 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
+      return _linalg_cond_exception_helper(self);
+    } else {
+      TORCH_CHECK(false, "linalg_cond got an unexpected error:\n", e.what());
+    }
+  }
+  std::array<int64_t, 2> dim_arr = {-2, -1};
+  optional<IntArrayRef> dim = IntArrayRef(dim_arr);
+  Tensor norm_self = at::linalg_norm(self, ord, dim);
+  Tensor norm_inverse = at::linalg_norm(self_inverse, ord, dim);
+  Tensor result = norm_self * norm_inverse;
+  return result;
+}
+
+Tensor& linalg_cond_out(Tensor& result, const Tensor& self, optional<Scalar> opt_ord) {
+  // If ord == None or ord == ±2 then SVD is used to compute the condition number
+  // the result is always real-valued, for other cases it is complex-valued for the complex-valued input.
+  ScalarType real_dtype = toValueType(typeMetaToScalarType(self.dtype()));
+  Scalar ord = opt_ord.has_value() ? opt_ord.value() : 2;
+  auto expected_dtype = std::abs(ord.toDouble()) == 2.0 ? real_dtype : self.scalar_type();
+
+  TORCH_CHECK(result.scalar_type() == expected_dtype,
+    "result dtype ", result.scalar_type(), " does not match the expected dtype ", expected_dtype);
+
+  Tensor result_tmp = at::linalg_cond(self, opt_ord);
+  at::native::resize_output(result, result_tmp.sizes());
+  result.copy_(result_tmp);
+  return result;
+}
+
+// Frobenius or nuclear norms
+Tensor linalg_cond(const Tensor& self, std::string ord) {
+  TORCH_CHECK(self.numel() > 0, "linalg_cond is not defined for empty tensors.");
+  // the same checks as squareCheckInputs(self) but with a slightly more informative error message
+  TORCH_CHECK(self.dim() >= 2, "linalg_cond only supports matrices or batches of matrices, but got a tensor with ",
+    self.dim(), " dimensions.");
+  TORCH_CHECK(self.size(-1) == self.size(-2),
+              "linalg_cond only supports square matrices or batches of square matrices "
+              "but got ", self.size(-1), " by ", self.size(-2), " matrices");
+
+  // Ignore errors if not invertible, result is INFINITY in this case
+  // Currently checking for error in at::inverse causes cross-device data movement
+  // For batched input if at least one matrix in the batch is not invertible,
+  // then the result for all other (possibly) invertible matrices will be infinity as well
+  // since there is currently no way to use at::inverse with silent errors
+  Tensor self_inverse;
+  try {
+    self_inverse = at::inverse(self);
+  } catch (const std::exception& e) {
+    if (strstr(e.what(), "singular")) {
+      return _linalg_cond_exception_helper(self);
+    } else {
+      TORCH_CHECK(false, "linalg_cond got an unexpected error:\n", e.what());
+    }
+  }
+  std::array<int64_t, 2> dim_arr = {-2, -1};
+  optional<IntArrayRef> dim = IntArrayRef(dim_arr);
+  Tensor norm_self = at::linalg_norm(self, ord, dim);
+  Tensor norm_inverse = at::linalg_norm(self_inverse, ord, dim);
+  Tensor result = norm_self * norm_inverse;
+  return result;
+}
+
+Tensor& linalg_cond_out(Tensor& result, const Tensor& self, std::string ord) {
```

Comment:
These two functions that take strings are very similar to the functions that take integers. What would you think of reusing the same function for both by using something like an enum to represent the norm values? Then the string signatures would just translate their inputs to the enum values and call the common function.

Reply:
I removed the duplicated code, now using `c10::variant<Scalar, std::string>` + `c10::visit`. Using enums in this case would require conversion functions `linalg_cond: scalar or string` -> `norm_ord_to_enum` -> helper function -> `enum_to_norm_ord` -> `linalg_norm`.

Reply:
The approach you took looks great.